g2p/mappings/utils.py

```diff
@@ -21,6 +21,7 @@
 
 GEN_DIR = os.path.join(os.path.dirname(langs.__file__), "generated")
 GEN_CONFIG = os.path.join(GEN_DIR, "config.yaml")
+IntOrNone = TypeVar("IntOrNone", bound=Union[int, None])
```

Comment:
Wow, no, it really isn't! I don't get it. I just tried, to see what would happen, and mypy complains all over!

Why does mypy accept an IntOrNone where it expects an int, but not an Optional[int] in the same situation? The very definition of Optional[X] is Union[X, None], which is what you bind IntOrNone to!

Reply:
It gets more confusion by the second. mypy won't let me assign a Union[int, None] to an int. What magic did you do to get it to let you assign an IntOrNone to an int?

Reply:
I just read up on what bound means in TypeVar, but I still don't get it.
Mypy even lets me do this:
```
v1: IntOrNone = None
v2: int = v1
```
this clearly assigns None to v2, which should not have been allowed. Is this a mypy bug?

Reply:
OK, now I get it. I've done some more narrow tests in isolation, and I see that it's for function arguments, and the narrower type gets bound are call time based on the caller's actual parameters. That make a lot more sense now.

Reply:
And when I was testing, I failed to notice that `v1: IntOrNone` produces its own error, complaining the type variable `IntOrNone` itself is unbound, and therefore no, it's not true mypy lets me do `v2: int = v1`, it's just that that line is not where the error is!

Reply:
Hi! Yes, it's all very confusing, but basically mypy doesn't consider `List[Tuple[int, int]]` to be a subtype of `List[Tuple[int, Optional[int]]]` for some reason that I don't entirely grasp (since it *does* allow you to assign a `Tuple[int, int]` to a `Tuple[int, Optional[int]]`).  Something about invariant variant covariant mumble mumble mumble.

While we might be able to get around this by using `Sequence` instead of `List`, the advantage of using a type variable here is that if you pass a `List[Tuple[int, int]]` to `compose_indices`, the return type will be `List[Tuple[int, int]]` instead of `List[Tuple[int, Optional[int]]]`, which is what we want in a lot of cases where we *know* there won't be a `None` there.

Reply:
I will rename `IntOrNone` to `IntOrOptionalInt` maybe ;-)

Reply:
to illustrate this, you can try running mypy on:

```python
from typing import Tuple, List, Sequence, Optional

a: Tuple[int, int] = (1, 2)
b: Tuple[int, Optional[int]] = (1, 2)
c: Tuple[int, Optional[int]] = a
# this would fail
# d: Tuple[int, int] = b


aa: List[Tuple[int, int]] = [a, a]
ab: List[Tuple[int, Optional[int]]] = [a, b]

# Sequence works here, List doesn't
aaaaaa: Sequence[Tuple[int, Optional[int]]] = aa
# this would fail
# aaaaaa: List[Tuple[int, Optional[int]]] = aa
```

Reply:
Thanks for this. Now that I understand what bound TypeVars actually, this makes a lot of sense. I find it really neat how they let you assign the output type to the input that, I can see that being really useful. And it's all still static-checkable since you get that input (and therefore output) type at each place when a function gets called.

I like the name IntOrOptionalInt, it makes the intent clear.